modules/040-control-plane-manager/images/control-plane-manager/controller/converge.go

```diff
@@ -0,0 +1,278 @@
+
+	if err := removeDirectory(dstDir); err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(dstDir, 0755); err != nil {
+		return err
+	}
+
+	dirEntries, err := os.ReadDir(configPath)
+	if err != nil {
+		return err
+	}
+
+	for _, entry := range dirEntries {
+		if entry.IsDir() {
+			continue
+		}
+		if !strings.HasPrefix(entry.Name(), "extra-file-") {
+			continue
+		}
+
+		if err := installFileIfChanged(filepath.Join(configPath, entry.Name()), filepath.Join(dstDir, strings.TrimPrefix(entry.Name(), "extra-file-")), 0644); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func convergeComponents() error {
+	log.Infof("phase: converge kubernetes components")
+	for _, v := range []string{"kube-apiserver", "kube-controller-manager", "kube-scheduler", "etcd"} {
+		if err := convergeComponent(v); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func convergeComponent(componentName string) error {
+	log.Infof("converge component %s", componentName)
+
+	//remove checksum patch, if it was left from previous run
+	_ = os.Remove(filepath.Join(deckhousePath, "kubeadm", "patches", componentName+"999checksum.yaml"))
+
+	if err := prepareConverge(componentName, true); err != nil {
+		return err
+	}
+
+	checksum, err := calculateChecksum(componentName)
+	if err != nil {
+		return err
+	}
+
+	recreateConfig := false
+	if _, err := os.Stat(filepath.Join(manifestsPath, componentName+".yaml")); err == nil {
+		equal, err := manifestChecksumIsEqual(componentName, checksum)
+		if err != nil {
+			return err
+		}
+		if !equal {
+			recreateConfig = true
+		}
+	} else {
+		recreateConfig = true
+	}
+
+	if recreateConfig {
+		log.Infof("generate new manifest for %s", componentName)
+		if err := backupFile(filepath.Join(manifestsPath, componentName+".yaml")); err != nil {
+			return err
+		}
+
+		if err := generateChecksumPatch(componentName, checksum); err != nil {
+			return err
+		}
+
+		_, err := os.Stat("/var/lib/etcd/member")
+		if componentName == "etcd" && err != nil {
+			if err := etcdJoinConverge(); err != nil {
+				return err
+			}
+		} else {
+			if err := prepareConverge(componentName, false); err != nil {
+				return err
+			}
+		}
+
+		_ = os.Remove(filepath.Join(deckhousePath, "kubeadm", "patches", componentName+"999checksum.yaml"))
+
+	} else {
+		log.Infof("skip manifest generation for component %s because checksum in manifest is up to date", componentName)
+	}
+
+	return waitPoidIsReady(componentName, checksum)
+}
+
+func prepareConverge(componentName string, isTemp bool) error {
+	args := []string{"init", "phase"}
+	if componentName == "etcd" {
+		args = append(args, "etcd", "local", "--config", deckhousePath+"/kubeadm/config.yaml")
+	} else {
+		args = append(args, "control-plane", strings.TrimPrefix(componentName, "kube-"), "--config", deckhousePath+"/kubeadm/config.yaml")
+	}
+	if isTemp {
+		args = append(args, "--rootfs", config.TmpPath)
+	}
+	c := exec.Command(kubeadm(), args...)
```

Comment:
could you add the comment with full kubeadm command, please?

Reply:
yes, done